kernel/src/process/signal/pauser.rs

```diff
@@ -112,159 +47,108 @@ impl Pauser {
     /// it will returns [`ETIME`].
     ///
     /// [`ETIME`]: crate::error::Errno::ETIME
-    pub fn pause_until_or_timeout<F, R>(self: &Arc<Self>, cond: F, timeout: &Duration) -> Result<R>
+    fn pause_until_or_timeout<F, R>(&self, mut cond: F, timeout: &Duration) -> Result<R>
     where
         F: FnMut() -> Option<R>,
     {
+        if *timeout == Duration::ZERO {
+            return cond()
+                .ok_or_else(|| Error::with_message(Errno::ETIME, "the time limit is reached"));
+        }
         self.do_pause(cond, Some(timeout))
     }
+}
 
-    fn do_pause<F, R>(self: &Arc<Self>, mut cond: F, timeout: Option<&Duration>) -> Result<R>
+impl internal::PauseInner for Waiter {
+    fn do_pause<F, R>(&self, mut cond: F, timeout: Option<&Duration>) -> Result<R>
     where
         F: FnMut() -> Option<R>,
     {
-        let current_thread = Thread::current();
-        let sig_queue_waiter =
-            SigObserverRegistrar::new(current_thread.as_ref(), self.sig_mask, self.clone());
+        if let Some(res) = cond() {
+            return Ok(res);
+        }
 
-        let cond = || {
-            if let Some(res) = cond() {
-                return Some(Ok(res));
+        let current_thread = self
+            .binding_task()
+            .data()
+            .downcast_ref::<Weak<Thread>>()
+            .and_then(|thread| thread.upgrade());
+
+        let Some(posix_thread) = current_thread
+            .as_ref()
+            .and_then(|thread| thread.as_posix_thread())
+        else {
+            if let Some(timeout) = timeout {
+                return self.wait_until_or_timeout(cond, timeout);
+            } else {
+                return self.wait_until_or_cancelled(cond, || Ok(()));
             }
+        };
 
-            if sig_queue_waiter.is_interrupted() {
-                return Some(Err(Error::with_message(
+        let cancel_cond = || {
+            posix_thread.set_signalled_waker(self.waker());
```

Comment:
This is because we want something like this:
```rust
loop {
    cond();

    set_signalled_waker();       // in `cancel_cond()`
    check_pending_signals()?;    // in `cancel_cond()`
    wait();
}
```

Note that nothing in `cond()` prevents us from calling `pause_until` again. So the correct place to call `set_signalled_waker` is just before `check_pending_signals` and `wait`, not before `cond`.

I'm not sure if we have existing code where `pause_until` is called in a nested way. However, I want to allow it because:
 - `wait_until` clearly allows nested calls. This will happen if we try to hold a mutex in the `cond` closure. So `pause_util` should allow this as well.
 - If we don't allow it, this will be an invariant that will be hard to enforce by users of the method, since `cond` can be quite complex.